### PR TITLE
Finally add relative location directions

### PR DESCRIPTION
**Pull request by TheLimeGlass**
### Description
Adds SkQuery's second main feature, relative locations between eachother. You can currently do this with vectors, but not the direction system. Adding this to Skript brings be sane again.

Cleaned up ExprDirection class too.

This allows for `push player in the direction of player to target block`

Note: I also attempted to apply `away` https://github.com/SkriptLang/Skript/issues/1671 but it's very hard as there is no defined relative reference when only defining `away`. I got it working for directions and blocks, but not entities. I'll maybe try another time, I don't know. My attempt was utilizing default expressions to bring in a reference like `[%location/direction/entity%]` but that can be used in very weird ways.

### CODE REVIEW

**Review by sovdeeth — APPROVED** [review]
Looks good, apart from the failing checks.
Also, quick question, why the use of CollectionUtils.array here? Just wondering if it's a standard that should be used, and if so, why.

**Comment by TheLimeGlass** [comment]
> Looks good, apart from the failing checks. Also, quick question, why the use of CollectionUtils.array here? Just wondering if it's a standard that should be used, and if so, why.

It looks cleaner, and was used by Njol for everything.

File: src/main/java/ch/njol/skript/expressions/ExprDirection.java [hunk 1]
```diff
@@ -124,89 +125,104 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
...
 	@Override
-	public String toString(final @Nullable Event e, final boolean debug) {
-		final Expression<?> relativeTo = this.relativeTo;
-		return (amount != null ? amount.toString(e, debug) + " meter(s) " : "") + (direction != null ? Direction.toString(direction) :
-				relativeTo != null ? " in " + (horizontal ? "horizontal " : "") + (facing ? "facing" : "direction") + " of " + relativeTo.toString(e, debug) :
-						(horizontal ? "horizontally " : "") + Direction.toString(0, yaw, 1));
+	public String toString(@Nullable Event event, boolean debug) {
+		return (amount != null ? amount.toString(event, debug) + " meter(s) " : "") +
+				(direction != null ? Direction.toString(direction) : relativeTo != null ? " in " +
+				(horizontal ? "horizontal " : "") +(facing ? "facing" : "direction") + " of " +
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/expressions/ExprDirection.java:224]: Would this be a good opportunity to clean up the giant mess of 6 nested ternary statements?(:?:?:?:)

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprDirection.java [hunk 2]
```diff
@@ -124,89 +125,104 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
...
 		if (direction != null) {
-			final Vector v = direction.clone().multiply(ln);
-			ExprDirection d = next;
-			while (d != null) {
-				final Number n2 = d.amount != null ? d.amount.getSingle(e) : 1;
-				if (n2 == null)
+			Vector vector = this.direction.clone().multiply(length);
+			ExprDirection addDirection = next;
+			while (addDirection != null) {
+				Number number2 = addDirection.amount != null ? addDirection.amount.getSingle(event) : 1;
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprDirection.java:149]: needs a better name than number 2
maybe something like delta length, or length modifier, or something like that.

File: src/main/java/ch/njol/skript/expressions/ExprDirection.java [hunk 3]
```diff
@@ -124,89 +125,104 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
...
+				addDirection = addDirection.next;
 			}
-			assert v != null;
-			return new Direction[] {new Direction(v)};
+			assert vector != null;
+			return CollectionUtils.array(new Direction(vector));
 		} else if (relativeTo != null) {
-			final Object o = relativeTo.getSingle(e);
-			if (o == null)
+			Object object = relativeTo.getSingle(event);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/expressions/ExprDirection.java:159]: needs a better name than object

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprDirection.java [hunk 4]
```diff
@@ -58,53 +56,56 @@
...
 						") [%-direction%]",
-				"[%-number% [(block|met(er|re))[s]]] in [the] (0¦direction|1¦horizontal direction|2¦facing|3¦horizontal facing) of %entity/block% (of|from|)",
-				"[%-number% [(block|met(er|re))[s]]] in %entity/block%'[s] (0¦direction|1¦horizontal direction|2¦facing|3¦horizontal facing) (of|from|)",
+				"[%-number% [(block|met(er|re))[s]]] in [the] (0¦direction|1¦horizontal direction|2¦facing|3¦horizontal facing) of %entity/block% [of|from]",
+				"[%-number% [(block|met(er|re))[s]]] in %entity/block%'[s] (0¦direction|1¦horizontal direction|2¦facing|3¦horizontal facing) [of|from]",
 				"[%-number% [(block|met(er|re))[s]]] (0¦in[ ]front [of]|0¦forward[s]|2¦behind|2¦backwards|[to the] (1¦right|-1¦left) [of])",
-				"[%-number% [(block|met(er|re))[s]]] horizontal[ly] (0¦in[ ]front [of]|0¦forward[s]|2¦behind|2¦backwards|to the (1¦right|-1¦left) [of])");
+				"[%-number% [(block|met(er|re))[s]]] horizontal[ly] (0¦in[ ]front [of]|0¦forward[s]|2¦behind|2¦backwards|to the (1¦right|-1¦left) [of])",
+				"[[[in] the] direction] (of|from) %location% to %location%");
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprDirection.java:84]: Switch to using `:` for parse marks

File: src/main/java/ch/njol/skript/expressions/ExprDirection.java [hunk 5]
```diff
@@ -124,89 +125,104 @@ public boolean init(final Expression<?>[] exprs, final int matchedPattern, final
...
-				v.add(d.direction.clone().multiply(n2.doubleValue()));
-				d = d.next;
+				assert addDirection.direction != null; // checked in init()
+				vector.add(addDirection.direction.clone().multiply(number2.doubleValue()));
+				addDirection = addDirection.next;
 			}
-			assert v != null;
-			return new Direction[] {new Direction(v)};
+			assert vector != null;
+			return CollectionUtils.array(new Direction(vector));
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprDirection.java:157]: Why did you replace the array creation with that?

*And everywhere else*

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]
Now that we support Java 17+, you could use its features to clean up this class further.
And add tests

File: src/main/java/ch/njol/skript/expressions/ExprDirection.java (hunk 1 above)
Comment (APickledWalrus) [reply src/main/java/ch/njol/skript/expressions/ExprDirection.java:224]: Yeah could use the SyntaxStringBuilder

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
I don't have much else to add on top what has been already said. Feel free to re-request me once those reviews have been addressed.

**Comment by sovdeeth** [comment]
Closing due to inactivity.

**Pull request closed without merging**